mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp

```diff
@@ -114,15 +145,63 @@ void PowOp::getCanonicalizationPatterns(RewritePatternSet &results,
   results.add<SqrtReciprocalOptimization>(context);
 }
 
-LogicalResult SelectOp::canonicalize(SelectOp op, PatternRewriter &rewriter) {
-  auto notOp = op.getPred().getDefiningOp<tosa::LogicalNotOp>();
-  if (!notOp)
-    return failure();
-  rewriter.modifyOpInPlace(op, [&]() {
-    op.getOperation()->setOperands(
-        {notOp.getInput1(), op.getOnFalse(), op.getOnTrue()});
-  });
-  return success();
+struct SelectLogicalNotOptimization : public OpRewritePattern<tosa::SelectOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(tosa::SelectOp op,
+                                PatternRewriter &rewriter) const override {
+    auto notOp = op.getPred().getDefiningOp<tosa::LogicalNotOp>();
+    if (!notOp)
+      return failure();
+    rewriter.modifyOpInPlace(op, [&]() {
+      op.getOperation()->setOperands(
+          {notOp.getInput1(), op.getOnFalse(), op.getOnTrue()});
+    });
+    return success();
+  }
+};
+
+// This pattern canonicalizaties the following pattern:
+// %0 = tosa.greater_equal(input, 0)
+// %1 = tosa.select(%0, input, 0)
+// to tosa.clamp{min = 0, max = max}(input)
+// This pattern occurs in decompositions of LeakyReLU/PReLU with an alpha
+// of zero
+struct SelectToClampOptimization : public OpRewritePattern<tosa::SelectOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(tosa::SelectOp op,
+                                PatternRewriter &rewriter) const override {
+    auto onFalse = op.getOnFalse();
+    DenseElementsAttr onFalseAttr;
+    if (!matchPattern(onFalse, m_Constant(&onFalseAttr)) ||
+        !isSplatZero(onFalse.getType().getElementType(), onFalseAttr)) {
+      return failure();
+    }
+    auto geq = op.getPred().getDefiningOp<tosa::GreaterEqualOp>();
+    if (!geq) {
+      return failure();
+    }
+    DenseElementsAttr geqIn2Attr;
+    if (!matchPattern(geq.getInput2(), m_Constant(&geqIn2Attr)) ||
+        !isSplatZero(geq.getInput2().getType().getElementType(), geqIn2Attr)) {
+      return failure();
+    }
+    auto input = op.getOnTrue();
+    if (input.getDefiningOp() != geq.getInput1().getDefiningOp()) {
+      return failure();
+    }
+    rewriter.replaceOpWithNewOp<tosa::ClampOp>(
+        op, op.getType(), input, rewriter.getI64IntegerAttr(0),
+        rewriter.getI64IntegerAttr(std::numeric_limits<int64_t>::max()),
+        rewriter.getF32FloatAttr(0.0f),
+        rewriter.getF32FloatAttr(std::numeric_limits<float>::max()));
```

Comment:
In the current implementation, you have checked this select (x>=0, x , 0) and you are setting the min attr for a clip. You can also check this select. select(x>=0, 0, x) [0 is the on true] and you can set the max.

I believe you can generalize this canonicalization to (x>=y, y, x) where y is a splat const and you have to track if y is on true or false condition and set max or min correspondingly.

Reply:
Done